### DISCUSSION TITLE
Importing external MPC library

### DISCUSSION THREAD

**Discussion by zafer4481** [bitcraze/discussions#50] [Q&A]
Hi I'm implementing an onboard MPC controller. For this I would like to make use of an external library to solve some optimization problems:

https://github.com/osqp/osqp

Does anyone one know how I could integrate this library with the crazyflie firmware? Such that I can make use of these solver functions in combination with the controller of the crazyflie (especially in stabilizer.c)

**Comment by krichardsson** [comment]
I think you can use ibdw1000 as a template, search for it in the source tree

1. Make sure the files are in the file tree. One option is to copy them into a directory somewhere, the other is to use git submodules (see https://git-scm.com/book/en/v2/Git-Tools-Submodules)
2. Make header files are available to the compiler by modifying https://github.com/bitcraze/crazyflie-firmware/blob/master/Makefile#L46
3. Tell Kbuild which files to compile by modifying or adding Kbuild files. If you add the lib to the `vendor` directory using git submodules, you should add the files in https://github.com/bitcraze/crazyflie-firmware/blob/master/vendor/Kbuild
